From ebbcb8e84ff40782f5a601ce66f7fe0c75005cd7 Mon Sep 17 00:00:00 2001
From: Janne Karhunen <Janne.Karhunen@gmail.com>
Date: Tue, 26 Oct 2021 11:19:12 +0300
Subject: [PATCH] arm64 softmmu: log translations when requested

Signed-off-by: Janne Karhunen <Janne.Karhunen@gmail.com>
---
 target/arm/tlb_helper.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/target/arm/tlb_helper.c b/target/arm/tlb_helper.c
index 3107f9823e..d5dc867966 100644
--- a/target/arm/tlb_helper.c
+++ b/target/arm/tlb_helper.c
@@ -155,6 +155,7 @@ bool arm_cpu_tlb_fill(CPUState *cs, vaddr address, int size,
 {
     ARMCPU *cpu = ARM_CPU(cs);
     ARMMMUFaultInfo fi = {};
+    int el, vmid;
 
 #ifdef CONFIG_USER_ONLY
     int flags = page_get_flags(useronly_clean_ptr(address));
@@ -185,6 +186,15 @@ bool arm_cpu_tlb_fill(CPUState *cs, vaddr address, int size,
                         core_to_arm_mmu_idx(&cpu->env, mmu_idx),
                         &phys_addr, &attrs, &prot, &page_size,
                         &fi, &cacheattrs);
+
+    el = arm_mmu_idx_to_el(mmu_idx | ARM_MMU_IDX_A);
+    vmid = cpu->env.cp15.vttbr_el2 >> 48;
+    qemu_log_mask(CPU_LOG_MMU, "%u %u %lx %lx %c%c%c\n",
+                  el, vmid, address, phys_addr,
+                  prot & PAGE_READ ? 'r' : '-',
+                  prot & PAGE_WRITE ? 'w' : '-',
+                  prot & PAGE_EXEC ? 'x' : '-');
+
     if (likely(!ret)) {
         /*
          * Map a single [sub]page. Regions smaller than our declared
-- 
2.25.1

